src/Phpro/SoapClient/Soap/Engine/Metadata/MetadataOptions.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Phpro\SoapClient\Soap\Engine\Metadata;
+
+use Phpro\SoapClient\Soap\Engine\Metadata\Manipulators\MethodsManipulatorInterface;
+use Phpro\SoapClient\Soap\Engine\Metadata\Manipulators\MethodsManipulatorChain;
+use Phpro\SoapClient\Soap\Engine\Metadata\Manipulators\TypesManipulatorInterface;
+use Phpro\SoapClient\Soap\Engine\Metadata\Manipulators\TypesManipulatorChain;
+
+final class MetadataOptions
+{
+    /**
+     * @var MethodsManipulatorInterface
+     */
+    private $methodsManipulator;
+
+    /**
+     * @var TypesManipulatorInterface
+     */
+    private $typesManipulator;
+
+    public function __construct()
```

Comment:
How about using DI principle here ?

Once it's done, we can get rid of the 2 `getters` as well. I always find it weird to use `getters` and `withers`.

Because of the `getters`, we could potentially taint the object properties and then, the immutability would be broken. 

```suggestion
    public function __construct(MethodManipulatorInterface $methodManipulator, TypeManipulatorInterface $typeManipulator)
```

Reply:
I made both possible by adding a static constructor.
In this case : it is more a configurable object which should be as flexible as possible IMO.
The manipulators should only have an `__invoke` method, so they should never be mutable.

Reply:
Ok fair enough.

I just find it odd to wrap `MethodManipulatorInterface` and `TypeManipulatorInterface` in such object. I made a POC locally and we could get rid of `MetadataOptions` class pretty easily.

That's just my 2 cents.

Reply:
Agree, it's a bit of a strange class... We could get rid of it by making the MetadataInterface explicit in the engine's factory classes? Or what did you have in mind?

Reply:
I would just get rid of it and use `MethodManipulatorInterface` and `TypeManipulatorInterface` at the place.

I don't think there is a need to wrap them in a class `MetadataOptions`, unless you plan to add more things to it later on, but I doubt.

Reply:
I played around with it a bit more and decided to leave it as-is for now.
The problem is that I want to specify the "defaults" somewhere, but don't want to change too many places at the same time.
This intermediate object does exactly that. Although I agree that at some point we might kick it out for direct injection of a metadata object. Time will tell ... :)